sympy/physics/control/tests/test_lti.py

```diff
@@ -0,0 +1,233 @@
+    assert tf.num == (s + 1)
+    assert tf.den == (s**2 + s + 1)
+    assert tf.args == (s + 1, s**2 + s + 1)
+
+    tf1 = TransferFunction(s + 4, s - 5)
+    assert tf1.num == (s + 4)
+    assert tf1.den == (s - 5)
+    assert tf1.args == (s + 4, s - 5)
+
+    # using different polynomial variables.
+    tf2 = TransferFunction(p + 3, p**2 - 9)
+    assert tf2.num == (p + 3)
+    assert tf2.den == (p**2 - 9)
+    assert tf2.args == (p + 3, p**2 - 9)
+
+    tf3 = TransferFunction(p**3 + 5*p**2 + 4, p**4 + 3*p + 1)
+    assert tf3.args == (p**3 + 5*p**2 + 4, p**4 + 3*p + 1)
+
+    # no pole-zero cancellation on its own.
+    tf4 = TransferFunction((s + 3)*(s - 1), (s - 1)*(s + 5))
+    assert tf4.den == (s**2 + 4*s - 5)
+    assert tf4.args == (s**2 + 2*s - 3, s**2 + 4*s - 5)
+
+    # purely symbolic polynomials.
+    tf5 = TransferFunction(p, p)
+    assert tf5.num == tf5.den == p
+    assert tf5.args == (p, p)
+
+    tf6 = TransferFunction(5, 6)
+    assert tf6.num == 5
+    assert tf6.den == 6
+    assert tf6.args == (5, 6)
+
+    # ValueError when denominator is zero.
+    raises(ValueError, lambda: TransferFunction(4, 0))
+    raises(ValueError, lambda: TransferFunction(s, 0))
+    raises(ValueError, lambda: TransferFunction(0, 0))
+
+
+def test_TransferFunction_functions():
+    # explicitly cancel poles and zeros.
+    tf0 = TransferFunction(s**5 + s**3 + s, s - s**2)
+    a = TransferFunction(-s**4 - s**2 - 1, s - 1)
+    assert tf0.cancel_poles_and_zeros() == a
+
+    tf1 = TransferFunction((p + 3)*(p - 1), (p - 1)*(p + 5))
+    b = TransferFunction(p + 3, p + 5)
+    assert tf1.cancel_poles_and_zeros() == b
+
+    # negation of TF.
+    tf2 = TransferFunction(s + 3, s**2 - s**3 + 9)
+    tf3 = TransferFunction(-3*p + 3, 1 - p)
+    assert tf2.__neg__() == TransferFunction(-s - 3, s**2 - s**3 + 9)
+    assert tf3.__neg__() == TransferFunction(3*p - 3, 1 - p)
+
+    # taking power of a TF.
+    tf4 = TransferFunction(p + 4, p - 3)
+    tf5 = TransferFunction(s**2 + 1, 1 - s)
+    result = TransferFunction(s**6 + 3*s**4 + 3*s**2 + 1, -s**3 + 3*s**2 - 3*s + 1)
+    assert tf4.pow(2) == TransferFunction(p**2 + 8*p + 16, p**2 - 6*p + 9)
+    assert tf5.pow(3) == result
+
+    # sympy's own functions.
+    assert tf3.xreplace({p: s}) == TransferFunction(-3*s + 3, 1 - s)
+    tf = TransferFunction(s - 1, s**2 - 2*s + 1)
+    assert factor(tf) == TransferFunction(s - 1, (s - 1)**2)
+
+
+def test_TransferFunction_addition():
+    tf1 = TransferFunction(s + 6, s - 5)
+    tf2 = TransferFunction(s + 3, s + 1)
+    tf3 = TransferFunction(s + 1, s**2 + s + 1)
+    tf4 = TransferFunction(p, 2 - p)
+    assert tf1.add(tf2) == TransferFunction(2*s**2 + 5*s - 9, s**2 - 4*s - 5)
+    assert tf3.add(tf4) == \
+        TransferFunction(p*s**2 + 2*s + 2, -p*s**2 - p*s - p**2 + 4*s + 2)
+    assert tf1.add(s - 1) == TransferFunction(s**2 - 5*s + 11, s - 5)
+    assert tf1.add(8) == TransferFunction(9*s - 34, s - 5)
```

Comment:
What is 'add()'? Combining two transfer functions in parallel?

Reply:
I think it's for combining them in parallel and Mul is for series.

@namannimmo10 What's the plan if the Functions are in a feedback loop
![image](https://user-images.githubusercontent.com/30147327/83147419-47404180-a115-11ea-968c-0eb44e4e479f.png)

Reply:
Will that work with expressions then?

```
tf1 = TransferFunction(...)
tf2 = TransferFunction(...)
tf3 = TransferFunction(...)
tf4 = TransferFunction(...)

tf1*tf2 + tf3/tf4
```

The order of operations would do the muls and divides first then the parallel.  So I think that's correct.

> @namannimmo10 What's the plan if the Functions are in a feedback loop

We will probably want a method for that. But things like this should work:

```
H1 = TransferFunction(...)
H2 = TransferFunction(...)
yd = Function('yd')(s)
y1 = Function('y')(s)
u1 = Function('u')(s)

loop = sm.Eq((yd-y1*H2)*H1, y1)
solve(loop, y1/yd)
```

Reply:
A common parallel structure is the PID controller. This should work:

```
P = kp
I = ki*TransferFunction(1, s, s)
D = kd*TransferFunction(s, 1, s)
controller = P + I + D  # parallel combination
closed_loop_tf = plant = TransferFunction(num, den, s)
plant.feedback(controller)
```